tests/optim/test_oss.py

```diff
@@ -334,3 +334,62 @@ def test_collect_shards():
     mp.spawn(
         run_test_collect_shards, args=(world_size, reference_rank), nprocs=world_size, join=True,
     )
+
+
+def run_test_multiple_groups(rank, world_size):
+    # Only work with the even ranks, to check that the global_rank indexing is properly used
+    dist_init(rank, world_size)
+    sub_group_ranks = [0, 2, 4]
+    pg = torch.distributed.new_group(ranks=sub_group_ranks, backend="gloo")
+    device = "cpu"
+    epochs, batch, input_width, hidden, target_width = 5, 3, 20, 10, 5
+    loss_fn = torch.nn.L1Loss().to(device)
+
+    def check(optimizer):
+        # Just run a couple of random epochs, check that the model is properly updated
+        for _ in range(epochs):
+            target = torch.rand((batch, target_width), device=device)
+            inputs = torch.rand((batch, input_width), device=device)
+
+            def closure():
+                optimizer.zero_grad()
+                output = model(inputs)
+                loss = loss_fn(output, target)
+                loss.backward()
+
+                return loss
+
+            original_loss = optimizer.step(closure=closure)
+            post_update_loss = loss_fn(model(inputs), target)
+
+            assert (
+                original_loss - post_update_loss
+            ).norm() > 1e-6, f"Model update failed - difference {(original_loss - post_update_loss).norm()}"
```

Comment:
Why take the norm? Why not just abs?

If you check against an exact value the test will catch even more errors.

I'm not sure this is a strong enough test of correctness.

If OSS is working correctly then all workers should have the same params after step. Why not verify that all workers compute the same norm for the params?

Reply:
> Why take the norm? Why not just abs?
I'm not sure why this is important ? it just meant "if they are too different", any norm would do

> If you check against an exact value the test will catch even more errors.
> 
> I'm not sure this is a strong enough test of correctness.
it does catch the error that the linked PR fixed, which was the initial point, so at least it's not useless. Now it's probably possible to do better

> If OSS is working correctly then all workers should have the same params after step. Why not verify that all workers compute the same norm for the params?
sounds like a good idea, I agree on that one